src/caffe/layers/detectnet_transform_layer.cpp

```diff
@@ -0,0 +1,929 @@
+
+  vector<Mat3v> outputImages(inputImages.size());
+  Blob<Dtype>& outputLabels = *top[1];
+
+  for (size_t iImage = 0; iImage != inputImages.size(); ++iImage) {
+    const Mat3v& inputImage = inputImages[iImage];
+    const vector<BboxLabel >& inputLabel = labels[iImage];
+
+    Mat3v& outputImage = outputImages[iImage];
+    Dtype* outputLabel = &outputLabels.mutable_cpu_data()[
+        outputLabels.offset(iImage, 0, 0, 0)
+    ];
+
+    transform(inputImage, inputLabel, outputImage, outputLabel);
+  }
+  // emplace images in output image blob:
+  matsToBlob(outputImages, *top[0]);
+}
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::transform(
+    const Mat3v& img,
+    const vector<BboxLabel >& bboxlist,
+    Mat3v& img_aug,
+    Dtype* transformed_label
+) {
+  uint32_t image_x = g_param_.image_size_x();
+  uint32_t image_y = g_param_.image_size_y();
+
+  //Perform mean subtraction on un-augmented image:
+  Mat3v img_temp = img.clone(); //size determined by scale
+  // incoming float images have values from 0..255, but OpenCV expects these
+  // values to be from 0..1:
+  img_temp /= Dtype(UINT8_MAX);
+  img_aug = Mat::zeros(image_y, image_x, img.type());
+  vector<BboxLabel > bboxlist_aug;
+  AugmentSelection as;
+  // We only do random transform as augmentation when training.
+  if (this->phase_ == TRAIN) {
+    //TODO: combine hueRotation and desaturation if performance is a concern,
+    // as we must redundantly convert to and from HSV colorspace for each
+    // operation
+    as.hue_rotation = augmentation_hueRotation(img_temp, img_temp);
+    as.saturation = augmentation_desaturation(img_temp, img_temp);
+    // mean subtraction must occur after color augmentations as colorshift
+    // outside of 0..1 invalidates scale
+    meanSubtract(img_temp);
+    // images now bounded from -0.5...0.5
+    as.scale = augmentation_scale(img_temp, img_aug, bboxlist, bboxlist_aug);
+    as.degree =
+        augmentation_rotate(img_aug, img_temp, bboxlist_aug, bboxlist_aug);
+    as.flip = augmentation_flip(img_temp, img_aug, bboxlist_aug, bboxlist_aug);
+    as.crop = augmentation_crop(img_aug, img_temp, bboxlist_aug, bboxlist_aug);
+    img_aug = img_temp;
+  }
+  else {
+    // perform mean subtraction:
+    meanSubtract(img_temp);
+
+    // deterministically scale the image and ground-truth, if requested:
+    transform_scale(
+        img_temp,
+        img_aug,
+        bboxlist,
+        bboxlist_aug,
+        Size(image_x, image_y)
+    );
+  }
+
+  CHECK_EQ(img_aug.cols, image_x);
+  CHECK_EQ(img_aug.rows, image_y);
+
+  // generate transformed_label based on bboxlist_aug
+  coverage_->generate(transformed_label, bboxlist_aug);
+
+  // visualize it: both before and after augmentation
+  if (g_param_.visualize()) {
+    visualize_bboxlist(
+        img,
+        img_aug,
+        bboxlist,
+        bboxlist_aug,
+        transformed_label,
+        as
+    );
+  }
```

Comment:
I asked @ssarathy about it and she said I should just leave it in. I think it's an awful hack and would be glad to remove it, but (as I said above) I didn't really want to start wading into cleaning up the code for real.